Digital Credentials: Add support for ISO/IEC TS 18013-7 (Annex c)

marcoscaceres opened the pull request:
#### 6dcf1e7e2a9f26587ec853af1ecd37b94583fa26
<pre>
Digital Credentials: Add support for ISO/IEC TS 18013-7 (Annex c)
<a href="https://bugs.webkit.org/show_bug.cgi?id=287593">https://bugs.webkit.org/show_bug.cgi?id=287593</a>
radar://143800384

Reviewed by Abrar Rahman Protyasha.

Adds support for the ISO/IEC TS 18013-7 (Annex c) format for the Digital Credentials API.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
(WebCore::CredentialRequestCoordinator::presentPicker):
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::jsToCredentialRequest):
(WebCore::DigitalCredential::discoverFromExternalSource):
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.idl:
* Source/WebCore/Modules/identity/DigitalCredentialsRequestData.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl:
* Source/WebCore/Modules/identity/MobileDocumentRequest.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialProtocol.h.
* Source/WebCore/Modules/identity/MobileDocumentRequest.idl: Copied from Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/290458@main">https://commits.webkit.org/290458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c43df0dfbbf0d0b5a3befb1fc6780ea3a052fdd3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45034 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95120 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40894 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10035 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17954 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69382 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26981 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93120 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7685 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81756 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49743 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36122 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40026 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77742 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37183 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96945 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12714 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78377 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77580 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77582 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19149 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22034 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20633 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10530 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17317 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22643 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17058 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20510 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18842 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/Modules/identity/DigitalCredential.cpp:
```diff
@@ -105,13 +130,25 @@ void DigitalCredential::discoverFromExternalSource(const Document& document, Cre
+    document.protectedFrame()
+        ->protectedPage()
+        ->credentialRequestCoordinator()
+        .presentPicker(WTFMove(promise), WTFMove(requestData), options.signal);
```
Hmm, we don't usually wrap chained calls like this. No strong preference though.

marcoscaceres replied:
Too jQuery? hehe...

aprotyas commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -53,12 +53,12 @@ class CredentialRequestCoordinator final : public RefCounted<CredentialRequestCo
     ~CredentialRequestCoordinator();
 
 private:
+    static bool canPresentDigitalCredentialsUI() { return m_hasDigitalCredentialsUI; }
```
Seems like an extra indirection, why not just define the static variable like this?

Also, you can make it `constexpr`!

```suggestion
    static constexpr bool canPresentDigitalCredentialsUI {
#if HAVE(DIGITAL_CREDENTIALS_UI)
        true
#else
        false
    };
```

And then you can use it like such:

```cpp
void CredentialRequestCoordinator::foobar()
{
    if (canPresentDigitalCredentialsUI) {
        ...
    }
}
```

aprotyas commented on Source/WebCore/Modules/identity/DigitalCredential.cpp:
```diff
@@ -98,20 +132,36 @@ void DigitalCredential::discoverFromExternalSource(const Document& document, Cre
+        std::visit([&](auto& credential) {
+            requestData.requests.append(credential);
+        },
+            credentialVariant);
```
Extra leading space
```suggestion
        credentialVariant);
```

marcoscaceres replied:
clang-tidy keeps moving it. I'll move the lambda to the next line so it looks better.

aprotyas commented on Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
```diff
@@ -26,13 +26,14 @@
 
 #include "IdentityCredentialProtocol.h"
-#include "OpenID4VPRequest.h"
+#include <JavaScriptCore/JSObject.h>
```
You can remove this and forward decl the `Object` class.

```
namespace JSC {
class JSObject;
}
```

aprotyas commented on Source/WebCore/Modules/identity/MobileDocumentRequest.h:
```diff
@@ -0,0 +1,36 @@
+ */
+
+#pragma once
+#include <wtf/text/WTFString.h>
```
Unnecessary include, please remove!

aprotyas reviewed (approved):
A few comments that I'd like addressing before the merge.

marcoscaceres committed:
Digital Credentials: Add support for ISO/IEC TS 18013-7 (Annex c)
https://bugs.webkit.org/show_bug.cgi?id=287593
radar://143800384

Reviewed by Abrar Rahman Protyasha.

Adds support for the ISO/IEC TS 18013-7 (Annex c) format for the Digital Credentials API.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
(WebCore::CredentialRequestCoordinator::presentPicker):
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::jsToCredentialRequest):
(WebCore::DigitalCredential::discoverFromExternalSource):
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.idl:
* Source/WebCore/Modules/identity/DigitalCredentialsRequestData.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl:
* Source/WebCore/Modules/identity/MobileDocumentRequest.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialProtocol.h.
* Source/WebCore/Modules/identity/MobileDocumentRequest.idl: Copied from Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/290458@main
